mimxrt: Add Quadrature Encoder and Pulse Counter classes.

robert-hh opened the pull request:
This PR adds Quadrature Encoder and Pulse Counter classes based on the Encoder hardware of the mimxrt MCUs. The base methods are as simple as possible, with options to make use of the hardware features supporting fast encoder sensors, like index pulses.

Tested with a slow manual encoder and a simulation of fast encoder/counter signals up to a input frequency of 25 MHz.

The PR is a re-submit of PR #7911, which could not be reopened. The conversation at that PR applies here as well (except for the UART part).

robert-hh committed:
mimxrt: Implement Quadrature Encoder and Counter classes.

These classes are based on the Quadrature Encoder blocks of the i.MXRT
MCUs.  The i.MXRT 102x has two encoders, the other ones four.  The i.MXRT
101x does not support this function.  It is implemented as two classes,
Encoder and Counter.

The number of pins that can be uses as inputs is limited by the MCU
architecture and the board schematics.  The Encoder class supports:
- Defining the module.
- Defining the input pins.
- Defining a pin for an index signal.
- Defining a pin for a reset signal.
- Defining an output pin showing the compare match signal.
- Setting the number of cycles per revolution (min/max).
- Setting the initial value for the position.
- Setting the counting direction.
- Setting a glitch filter.
- Defining callbacks for getting to a specific position, overrun and
  underrun (starting the next revolution).  These callbacks can be hard
  interrupts to ensure short latency.

The encoder counts all phases of a cycle.  The span for the position is
2**32, for the revolution is 2**16.  The highest input frequency is
CPU-Clock/24.  Note that the "phases" argument is emulated at the API
level (the hardware will always count all phases).

The Counter mode counts single pulses on input A of the Encoder.  The
configuration supports:
- Defining the module.
- Defining the input pin.
- Defining the counting direction, either fixed or controlled by the level
  of an input pin.
- Defining a pin for an index signal.
- Defining an output pin showing the compare match signal.
- Setting the counter value.
- Setting the glitch filter.
- Defining a callback which is called at a certain value.
- Settings for MIMXRT1015. The MIMXRT1015 MCU has only one encoder/counter
  unit.

The counting range is 0 - 2**32-1 and a 16 bit overrun counter.  The
highest input frequency is CPU-Clock/12.

The implementation of the `.irq()` method uses the common code from
`shared/runtime/mpirq.c`, including the `irq().flags()` and
`irq().trigger()` methods.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh committed:
docs: Add documentation for the mimxrt Encoder/Counter class.

This adds MIMXRT-specific arguments and methods, as a superset of the
original Encoder/Counter documentation.

The mimxrt pinout and quickref docs are updated with relevant information.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
@dpgeorge @projectgus   Now that the Encoder/Counter classes are available for ESP32, I have rebased and retested the implementation for MIMXRT. Since the Encoder/Counter hardware differs between ESP32 and MIMXRT, the ESP32 oriented test in extmod_hardware show quite a few deviations. Major reasons:

- The MIMXRT Encoder always counts every transition.  So count numbers are by a factor of 4 higher than in the ESP32. But it does not have the ESP32 glitch to need 3 transitions for the first count.
- Encoder/Counter input values cannot be read back using the machine.Pin methods, causing the connection test to fail.
- The MIMXRT Counter always counts the rising edge. The active edge cannot be changed.

projectgus commented:
Thanks @robert-hh for updating and noting these differences. (I think it's a good example of the potential for tests to find these hardware-specific parts as well.)

> The MIMXRT Encoder always counts every transition. So count numbers are by a factor of 4 higher than in the ESP32. But it does not have the ESP32 glitch to need 3 transitions for the first count.

The ESP32 version of this class has a documented port-specific `phases` parameter, that defaults to `1` but it can also count each transition by setting it `4`. So we could change the default value of `phases`, but another option would be to emulate the `phases` parameter in software on the mimxrt version. What do you think?

> Encoder/Counter input values cannot be read back using the machine.Pin methods, causing the connection test to fail.
> The MIMXRT Counter always counts the rising edge. The active edge cannot be changed.

I think the solution for these will probably be `@unittest.skipIf()` for the respective tests. Possibly in the top block we can set variables like `COUNT_FALLING = True` and then evaluate those in the skip clauses.

I am a bit surprised that `machine.Pin()` doesn't work for the input values, though. Is that because there are pins which can never be configured as GPIO, or is that a limitation that once the port muxes them to be counter inputs it doesn't/can't mux them back?

robert-hh commented:
Good day @projectgus, thank you for the swift reply.
Emulating Phase would be possible to a certain extend by dividing the MIMXRT counter by 2 or 4. That would however not cater a possible situation, that ESP32 increases the count with a specific transition. That seems to be the case and may be the reason for the first count in the test needing only 3 transitions. Maybe that could be solved by finding the proper start conditions of the inputs. Another indication of changing the count at a specific transition is the fact, that for ESP32 a single transition down from 0 decreases the counter to -1. A simply division would not cater for that.

Obviously one can skip the test for the FALLING edge of the counter. 

I did not look in detail why the Pin value() cannot read back when the Pin was configured for Encoder/Counter. Probably it's because for Encoder/Counter a PIN is configured for a different ALT mode. So maybe the test has to be changed to configure the Pin for machine.Pin mode first, make the connectivity test and then configure it as Encoder/Counter. I wondered why there is the connectivity test at all. The other tests to not have it.

robert-hh committed:
tests/extmod_hardware/machine_encoder.py: Add a MIMXRT configuration.

For Teensy 4.x.  The connectivity tests and the falling edge of the counter
test are skipped.

Signed-off-by: robert-hh <robert@hammelrath.com>

projectgus commented:
Sorry I didn't reply last week, @robert-hh - the updated PR looks great to me, thanks for accommodating the limits in the API and tests! I don't think I have any mimxrt board set up to do any local tests on, unfortunately (there might be a Teensy 4 hiding somewhere, will look later).

> I wondered why there is the connectivity test at all. The other tests to not have it.

I borrowed this from Ihor's test PR, but it's something I support building into tests - especially hardware-in-loop tests. The idea is to have a quick and user-friendly way to differentiate "the test environment is broken" from "the driver is broken". More than once I've had some unusual pattern of failures and spent too long debugging them before realising a wire is loose, or some equivalent prerequisite is missing (i.e with MicroPython, forgetting to do `mpremote rtc --set` after hard reset has gotten me more than once).

As you observe, MicroPython's tests don't have much of this kind of thing (yet) - but I think that's to the project's detriment. It costs time even for core developers (like me) who run the test suite a lot, but for developers who aren't already *very* familiar with the project then it's even more limiting. We get a lot of PRs from new(ish) contributors who are competent developers, but the relevant tests either haven't been run, or they ran and failed and the developer hasn't dug in to figure out why.

(Sorry for dropping in this somewhat tangential rant. cc @hmaerki and @dpgeorge as I know they've been thinking about tests an awful lot, lately.)

In the particular case of the connections test for Counter/Encoder, I think skipping them on mimxrt - as you've done here - is fine for now. Long term, putting them in a separate TestCase at the top of the same file (so they're run without the setUp step that inits the Encoder and muxes the pins) will probably allow them to pass on more platforms, but I can look at that in a follow-up if you like.

dpgeorge commented:
> More than once I've had some unusual pattern of failures and spent too long debugging them before realising a wire is loose, or some equivalent prerequisite is missing (i.e with MicroPython, forgetting to do `mpremote rtc --set` after hard reset has gotten me more than once).

That's what #16112 is trying to address, although it's gotten pretty lost among all the other PRs.  Might be nice to get that merged, I use it all the time now.

> The idea is to have a quick and user-friendly way to differentiate "the test environment is broken" from "the driver is broken

We are definitely missing a simple `Pin` test that toggles IO and checks it works.  That could run before any of these other more sophisticated tests to verify the hardware is wired up correctly.

But also, I think we need to break out all this board wiring configuration into separate files, one for each board, that can be reused in all hardware-related tests.  That's something I'm working on now.

dpgeorge commented:
@robert-hh thanks for keeping this PR alive.  It's looking very good, especially now that the existing tests only have minor modifications to work on the mimxrt port.

Regarding the docs: it seems you've added custom documentation for the mimxrt version of Counter/Encoder.  But there already exists generic documentation in `docs/library/machine.Counter.rst` and `docs/library/machine.Encoder.rst`.  Ideally we don't need to add any docs for mimxrt, it should just behave the same as the existing documentation describes.

My suggestion:
- keep the existing quickref that you have in this PR for mimxrt, that's useful
- move the pinout information from `docs/library/mimxrt.Encoder.rst` to `docs/mimxrt/pinout.rst`
- add any mimxrt specific things to the existing `machine.Counter.rst` and `machine.Encoder.rst` files
- remove the `mimxrt.Encoder.rst` file added in this PR (hopefully all its relevant information was put elsewhere by the above two points)

dpgeorge committed:
tests/extmod_hardware/machine_encoder.py: Separate the connection test.

Because it requires a different configuration of the pins (in `setUp`).
Eg on mimxrt pins used for an Encoder cannot be read.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
> In the particular case of the connections test for Counter/Encoder, I think skipping them on mimxrt - as you've done here - is fine for now. Long term, putting them in a separate TestCase at the top of the same file (so they're run without the setUp step that inits the Encoder and muxes the pins) will probably allow them to pass on more platforms

I have push a commit to this PR to do that, to have a separate unittest class that does the connection test.  Tested on TEENSY40 and ESP32_GENERIC_P4, and it works.

dpgeorge committed:
tests/extmod_hardware/machine_counter.py: Separate the connection test.

Because it requires a different configuration of the pins (in `setUp`).
Eg on mimxrt pins used for a `machine.Counter` cannot be read.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented on tests/target_wiring/mimxrt.py:
```diff
@@ -5,3 +5,7 @@
+
+encoder_loopback_id = 0
+encoder_loopback_out_pins = ("D2", "D5")
+encoder_loopback_in_pins = ("D3", "D4")
```
Can these pins be changed to:
```py
encoder_loopback_out_pins = ("D0", "D2")
encoder_loopback_in_pins = ("D1", "D3")
```
So that the UART loopback pins are the same as one of the encoder loopback pairs?  That would mean you only need two connections on the board rather than three.

Similarly, for the counter test, would be good to reuse the D0/D1 loopback pair.

robert-hh replied:
Yes. Works & is done.

robert-hh committed:
mimxrt/boards/make-pins.py: Add the XBAR tag to the AF names.

Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
@dpgeorge  I have rearranged the commits a little bit, and more squashing can be done to the test-related commits and the two about documentation.

dpgeorge commented:
Thanks @robert-hh for updating the docs.  It's now much clearer how the mimxrt code matches the existing docs (and esp32 implementation) and what's been added.

So there's now two things to do here:
1. Review the mimxrt implementation to see that the existing documented functionality works as expected.  It seems that is indeed the case because the tests pass. 
2. Review and discuss the additions to the Counter and Encoder classes.  I can see you've been quite thorough there, adding support for the full mimxrt behaviour (or most of it).

Ideally (1) and (2) would be separate PRs, because (1) is much easier to review and merge, whereas (2) may take some back and forth to finalise the enhanced API.  But I can understand that might be more effort than you have time for.  So it's up to you how to proceed, ether split this PR into two where the first just implements existing documented behaviour, or keep it as one PR and it'll take a bit more time to review.

robert-hh commented:
Thank you for the comment. The decision is hard to make, because the underlying hardware between MIMXRT and ESP32 is different. Of course I can reduce behavior and the set of options and methods to being as close as possible to the ESP's PCNT/Encoder/Counter triple. But doing that by removing code and putting it back in later seems error prone. I would prefer leaving the extra code in and just disabling it with a compile switch. 
So what would it be: 
1. features not available in the ESP32 version.
- defining the pins for the index, home and match signals. That is kind of independent and easily disabled and enabled. Maybe one of index or home can be dropped totally.
- the signed/unsigned option. We could go for signed only.
- the counter/cycles model. I could hide the cycles property and just set the count range to 32 bit at phase=4. Since the hardware always counts 4 phases, at a phases=1 setting the range would drop to 30 bit.
- the `id()` method. That one may be obsolete, once the `irq().flags()` method is available.
2. Feature that are slightly different.
- the `edge` option for ESP32's counter, related to PCNT's rising/falling options. That is not supported for MIMXRT, and I did not find a way to implement it. Even if the chip has a configurable logic block, that can be set as inverter, there seems to be no way to route the output of that block to a Encoder/Counter input,.
- a Pin defining the count direction.  That feature is available on MIMXRT, and available through the mode_pin option of the ESP32 PCNT class. Only the way to specify it is different.
- The IRQ events. The only common event is a counter match, which is called IRQ_MATCH for MIMXRT and IRQ_THRESHOLD0 (or 1) for ESP32. I could call it IRQ_THRESHOLD0 at the MIMXRT port. But the other events are different and I see no way to emulate e.g. ESP32's events properly in the MIMXRT port or the opposite.
- Encoder. status(), Counter..status() vs irq.flags(). That can be harmonized to provide the irq()'s flag method.
- Reading the MIMXRT data sheets again, it seems that the `min` and `max` options of the ESP32.PCNT class could be enabled.


P.S.: Being a pensioner, I should have every time of the world. But since everyone assumes this (and asks me to do this or that), it's way less. Anyhow, time is not a problem.

robert-hh commented:
I made the suggested changes.

1. `cpc` was renamed to `max`, and a `min` option is added. Note that these behave slightly different to ESP32, in that they define the lower and upper limit for a modulus style counting. While ESP32 resets the counter to 0, MIMXRT rolls over or under. In any case, the event can cause an interrupt.
2. The `iqr()` implementation is changed into using the common code from runtime/mpirq. As suggested, I added a encoder.match()  method to set the counter match value. The ESP32 implementation of PCNT uses a keyword option of init() instead. That can be done here as well. The `id()` and `status()` methods are removed.

The names of the trigger events are still not changed. They could change to:

IRQ_MATCH  into IRQ_THRESHOLD0
IRQ_ROLL_UNDER into IRQ_MIN
IRQ_ROLL_OVER into IRQ_MAX.

Another change is the behavior of the index and home pins. They do now different things. The index pin transition now just increments or decrements the cycles counter. The home pin transition just resets the pulse counter.

The documentation is not yet updated. That will be done when the code is settled.

dpgeorge commented on docs/library/machine.Counter.rst:
```diff
@@ -47,12 +47,49 @@ Methods
-     or equal to this value. The default is 0 (no filter). *(Supported on ESP32)*
+     or equal to this value. The default is 0 (no filter). *(Supported on ESP32 and MIMXRT)*
+ 
+   - *max*\=value. Specify the upper counting range. The counter will count up
```
This style of using `=value` is not used elsewhere in our docs, so I suggest not using it here.  Instead just do:
```
*max* specifies the upper counting range. The counter ...
```
Also worth specifying the default value.

The same comments apply to all the other keyword arguments added in this PR.

dpgeorge reviewed (commented):
@robert-hh I've now done my final, full review of this PR.  Once the above comments are resolved, this should be good to merge!

robert-hh commented on docs/mimxrt/quickref.rst:
```diff
@@ -556,6 +556,61 @@ port and LAN(1) for the 1G port.
+
+    qe = Encoder(0, Pin("D0"), Pin("D1")) # create Quadrature Encoder object
+    qe.value()                      # get current counter values
+    qe.value(0)                     # set value and cycles to 0
```
Changed.

dpgeorge commented on docs/library/machine.Counter.rst:
```diff
@@ -79,15 +116,66 @@ Methods
+
+   The callback is called, when the Counter value matches the value defined by the
+   *match* keyword option of the constructor or the :meth:`Counter.init()` method. For fast signals,
+   the actual counter value may be different from the trigger value.
```
I think these 3 lines should be simplified and go in the bullet-point list above, in the `Counter.IRQ_MATCH` line.  Because they are only relevant for `IRQ_MATCH`.

robert-hh commented on docs/library/machine.Encoder.rst:
```diff
@@ -70,3 +104,59 @@ Methods
+
+   The callback is called, when the Counter value matches the value defined by the
+   *match* keyword option of the constructor or the :meth:`Counter.init()` method. For fast signals,
+   the actual counter value may be different from the trigger value.
```
Updated. I found two more places to change `Counter` -> `Encoder`

dpgeorge replied:
Great!

robert-hh commented on docs/mimxrt/quickref.rst:
```diff
@@ -556,6 +556,61 @@ port and LAN(1) for the 1G port.
+    counter.init(max=128)                 # set the upper counting range
+    counter.deinit()                      # turn off the Counter
+    counter.init(match=1000)              # create a match event at count 1000
+    counter.irq(handler=handler, Counter.IRQ_MATCH)
```
Updated.

dpgeorge reviewed (approved):
Thanks for updating, this looks good now.  Let's get it in so people can use it!

I will squash commits as part of the merge.

dpgeorge commented:
> The documentation is hard to get perfect.

It doesn't need to be perfect :smile:

robert-hh commented:
The files `mpconfigboard.h` for `ADAFRUIT_METRO_M7` and `MAKERDIARY_RT1011_NANO_KIT` are updated. The firmware builds and runs. To be sure, I built the binaries for all MIMXRT boards, and all build fine. The change is published as PR #18908.

For history: Since the MIMXRT1010 does not support Encoder, these boards were not tested with machine.Encoder, and being introduced later that the first versions of the PR12347, they were not any more considered as relevant. So building all boards is a required test step.
